### DISCUSSION TITLE
Import issue where file 2 is not running

### DISCUSSION THREAD

**Discussion by apacheli** [denoland/deno#9082] [Q&A]
Oddly, I seem to be getting a "not defined error" despite having the code correct. I would assume that this is likely a compile issue as I was able to replicate it almost instantly.
```ts
// test1.ts
import { A } from "./test2.ts";

console.log("test 1");

console.log(A.B);
```
```ts
// test2.ts
export const enum A {
  B
}

console.log("Is this even running");
```
Running these examples yields these results:
![image](https://user-images.githubusercontent.com/43933794/104133541-d408da00-5349-11eb-9741-b2eb5907d926.png)
From the logs, it seems like `test2.ts` doesn't run at all. The runtime seems to die here.

(If needed, I can move this to the issue tracker, which I probably should've done at first.)

**Comment by apacheli** [comment]
Update:

Importing `test2.ts` without importing any of its exports logs this:
```ts
import { A } from "./test2.ts";
import "./test2.ts";
```
![image](https://user-images.githubusercontent.com/43933794/104134181-b3428380-534d-11eb-927f-5b2e06c02e46.png)
It *is* running, though the export still doesn't happen.

I tried again with `import` as a function, and that works. However, I don't think that this is the solution.
```ts
const { A } = await import("./test2.ts");
```
![image](https://user-images.githubusercontent.com/43933794/104134248-0fa5a300-534e-11eb-9fbd-0bb3e1d6692e.png)

I also checked to see if `const enum` was the problem, and it is not.

**Comment by kitsonk** [comment]
`const enum` are not supported in Deno.  We need to make that clearer in the documentation.  We only support TypeScript in the `isolatedModules` mode, which is enabled by default.  I has assumed that `isolateModules` would report errors with `const enum` but it appears it doesn't.  `const enum` aren't ever exported, because they are fully erased from the emit and the emitted code ends up with a "hardcoded" value.  Because it is a type directed emit (e.g. what the type information is impacts what the emitted code is, which is really a TypeScript anti-pattern) they are not supported in Deno.  TypeScript/`tsc` continues to support it, though it runs afoul of the TypeScript design principles, because they don't want to break backwards compatibility.

**Reply by apacheli** [reply]
> 
> 
> `const enum` are not supported in Deno. We need to make that clearer in the documentation. We only support TypeScript in the `isolatedModules` mode, which is enabled by default. I has assumed that `isolateModules` would report errors with `const enum` but it appears it doesn't. `const enum` aren't ever exported, because they are fully erased from the emit and the emitted code ends up with a "hardcoded" value. Because it is a type directed emit (e.g. what the type information is impacts what the emitted code is, which is really a TypeScript anti-pattern) they are not supported in Deno. TypeScript/`tsc` continues to support it, though it runs afoul of the TypeScript design principles, because they don't want to break backwards compatibility.

Removing `const enum` seemed to have solved the issue for the most part, but I'm still confused on why the compiled code seems to be inconsistent whether I include `const` or not. I checked the compiled code in `%appdata%/local/deno` and it doesn't include any `import` statements:
```js
// These 2 results are without `const enum` in test2.ts
// test1.ts.js
console.log(A.B);
export {};

// test2.ts.js
export var A;
(function (A) {
    A[A["B"] = 10] = "B";
})(A || (A = {}));
```

It seems like Deno is ignoring the fact that `test1.js` is not being edited, therefore it doesn't re-compile?

This is what I did to replicate the error:
- Create both files with the initial code samples
- Expect error when running `test1.ts`, intended yes because `isolatedModules`
- Remove `const` from enum in `test2.ts`, and run `test1.ts` again. This also errors, which is probably unintended?
- Edit `test1.ts` however you wish, run again, and it doesn't error

**Comment by nayeemrmn** [comment]
This is a tsc transpilation bug tracked in https://github.com/microsoft/TypeScript/issues/40499, #7605.
